ports/nrf/common-hal/busio/I2C.c

```diff
@@ -79,20 +79,47 @@ void common_hal_busio_i2c_never_reset(busio_i2c_obj_t *self) {
     }
 }
 
+static nrfx_err_t _safe_twim_enable(busio_i2c_obj_t *self) {
+    // check to see if bus is in sensible state before enabling twim
+    nrfx_err_t recover_result;
+    nrf_gpio_cfg_input(self->scl_pin_number, NRF_GPIO_PIN_PULLDOWN);
+    nrf_gpio_cfg_input(self->sda_pin_number, NRF_GPIO_PIN_PULLDOWN);
+
+    common_hal_mcu_delay_us(10);
+
+    nrf_gpio_cfg_input(self->scl_pin_number, NRF_GPIO_PIN_NOPULL);
+    nrf_gpio_cfg_input(self->sda_pin_number, NRF_GPIO_PIN_NOPULL);
+
+    // We must pull up within 3us to achieve 400khz.
+    common_hal_mcu_delay_us(3);
+
+    if (!nrf_gpio_pin_read(self->sda_pin_number) || !nrf_gpio_pin_read(self->scl_pin_number)) {
+        // bus not in a sane state - try to recover
+        recover_result = nrfx_twim_bus_recover(self->scl_pin_number, self->sda_pin_number);
+        if (NRFX_SUCCESS != recover_result) {
+            // return error message if unable to recover the bus
+            return recover_result;
+        }
+    }
+
+    nrfx_twim_enable(&self->twim_peripheral->twim);
+    return NRFX_SUCCESS;
+}
+
 static uint8_t twi_error_to_mp(const nrfx_err_t err) {
     switch (err) {
         case NRFX_ERROR_DRV_TWI_ERR_ANACK:
             return MP_ENODEV;
         case NRFX_ERROR_BUSY:
             return MP_EBUSY;
+        case NRFX_SUCCESS:
+            return 0;
         case NRFX_ERROR_DRV_TWI_ERR_DNACK:
         case NRFX_ERROR_INVALID_ADDR:
-            return MP_EIO;
+        case NRFX_ERROR_INTERNAL:
```

Comment:
Since these all return `MP_EIO`, you could just remove them and have them be under `default:`.

Reply:
I think I'd disagree here. I'd quite like having the expected errors made explicit, so we already know that e.g. `NRFX_ERROR_DRV_TWI_ERR_DNACK` should deliberately be interpreted as `MP_EIO`. This means that in future if someone does something that will generate that error message, they know that is the correct translation. I would expect gcc to optimize this out in any event

Reply:
That is fine with me. I wasn't sure whether all possible errors were handled here.